### PR TITLE
Implement CLONE and DELETE

### PR DESCRIPTION
**Pull request by rogerl50**

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3263/pr-test (or any other room on that server)

**Comment by rogerl50** [comment]
See first commit note above for details of the syntax. A small example room is at A small example is at http://212.47.248.129:3261/clones

**Comment by 96LawDawg** [comment]
Testing of the clone button didn't work.  I tried to debug, but couldn't figure out what was going on.  I think it is the "applyVariables" part of the CLONE function, but not sure.

**Comment by rogerl50** [comment]
> Testing of the clone button didn't work. I tried to debug, but couldn't figure out what was going on. I think it is the "applyVariables" part of the CLONE function, but not sure.

That is sad. It runs properly in my debug setup. I don't know how to figure out what is wrong; maybe someone else can help.

**Comment by rogerl50** [comment]
> Testing of the clone button didn't work. I tried to debug, but couldn't figure out what was going on. I think it is the "applyVariables" part of the CLONE function, but not sure.

My bad. See updated URL above.

**Comment by 96LawDawg** [comment]
Okay, now it sort of worked.  Your test room, which is now completely broken, added 1 holder and 1 spinner when I pressed the Add button you created.  Pressing the Add button additional times did not, however, clone more.  The delete function worked on the one clone.  I tried to get it to create more than 1.  I changed the "count" parameter in the CLONE function.  No joy.  I changed the select from "set" to "add."  No joy.  I made the select choose every holder and spinner in the room.  That worked sort of.  I pressed the Add button a few times and a few holders and spinners got added.  I thought it might double the number every time.  It didn't.  It just added one at a time.  After about 5 presses, the room crashed and broke.

**Comment by rogerl50** [comment]
> Okay, now it sort of worked. Your test room, which is now completely broken, added 1 holder and 1 spinner when I pressed the Add button you created. Pressing the Add button additional times did not, however, clone more. The delete function worked on the one clone. I tried to get it to create more than 1. I changed the "count" parameter in the CLONE function. No joy. I changed the select from "set" to "add." No joy. I made the select choose every holder and spinner in the room. That worked sort of. I pressed the Add button a few times and a few holders and spinners got added. I thought it might double the number every time. It didn't. It just added one at a time. After about 5 presses, the room crashed and broke.

Every time you clone the original widget, the clones are placed in the same location, and they are all placed on top of each other. The xOffset and yOffset parameters control the offset from the *original* widget, not between each cloned object. So when you clicked Add once, you got three clones, all in the same place. When you clicked it again, you got three more, on top of the first three.

If you pressed Add "a few" times, say 10, there were 60 clones there. If you modified the button to clone everything and pressed that 5 times, you were up to 600 clones. I'm surprised that the room broke so badly, and it may be that there is a bug there.

**Comment by 96LawDawg** [comment]
Wait, you mean VTT can't handle 15,000 widgets in a room?  What a garbage program!

Okay, I'm a little bit of an idiot because I didn't recognize the clones were on top of each other.  But it takes "special" people like me to break things in stupid ways.  While the code works, we should at least include a warning in the wiki about people like me.  I mean, behavior like I accidentally tried, so a room doesn't get overwhelmed.

I would approve this, because it does work, but I want to see if there are any other comments about the functionality or features.

**Comment by rogerl50** [comment]
> > Okay, now it sort of worked. Your test room, which is now completely broken, added 1 holder and 1 spinner when I pressed the Add button you created. Pressing the Add button additional times did not, however, clone more. The delete function worked on the one clone. I tried to get it to create more than 1. I changed the "count" parameter in the CLONE function. No joy. I changed the select from "set" to "add." No joy. I made the select choose every holder and spinner in the room. That worked sort of. I pressed the Add button a few times and a few holders and spinners got added. I thought it might double the number every time. It didn't. It just added one at a time. After about 5 presses, the room crashed and broke.
> 
> Every time you clone the original widget, the clones are placed in the same location, and they are all placed on top of each other. The xOffset and yOffset parameters control the offset from the _original_ widget, not between each cloned object. So when you clicked Add once, you got three clones, all in the same place. When you clicked it again, you got three more, on top of the first three.
> 
> If you pressed Add "a few" times, say 10, there were 60 clones there. If you modified the button to clone everything and pressed that 5 times, you were up to 600 clones. I'm surprised that the room broke so badly, and it may be that there is a bug there.
> 
> The pr-test room on that port contains a currently-working version.



> Wait, you mean VTT can't handle 15,000 widgets in a room? What a garbage program!
> 
> Okay, I'm a little bit of an idiot because I didn't recognize the clones were on top of each other. But it takes "special" people like me to break things in stupid ways. While the code works, we should at least include a warning in the wiki about people like me. I mean, behavior like I accidentally tried, so a room doesn't get overwhelmed.
> 
> I would approve this, because it does work, but I want to see if there are any other comments about the functionality or features.

Agreed, the Wiki text needs to expand on the pretty bare-bones description I gave.

And I completely agree, we need to wait for a) a review by someone who has actually written JavaScript code before, and b) to see if there are other functionality comments.

File: client/js/widgets/button.js
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:126]: I would do `delete clone.id` here but it probably doesn't make a difference.

**Review by ArnoldSmith86 — APPROVED** [review]
I didn't test anything but I couldn't find problems with the code.

I would suggest that we discuss this PR in the meeting because it's major functionality.

File: client/js/widgets/button.js
Comment (rogerl50) [reply client/js/widgets/button.js:126]: The test later is `if (!widget.id)`, so it should behave the same.

**Review by RaphaelAlvez — CHANGES_REQUESTED** [review]
I can't figure out how to create a clone. Tried using you example and it didn't work, then I tried to do it myself and same result

**Review by RaphaelAlvez — APPROVED** [review]
I'm sorry for my last post. on the pr-test room it's working. Someone must have messed up the clone room

**Comment by rogerl50** [comment]
> I'm sorry for my last post. on the pr-test room it's working. Someone must have messed up the clone room

Yes, there are over 10000 widgets in that room from excessive cloning, it's useless.

**Comment by RaphaelAlvez** [comment]
> > I'm sorry for my last post. on the pr-test room it's working. Someone must have messed up the clone room
> 
> Yes, there are over 10000 widgets in that room from excessive cloning, it's useless.

OK so this PR is approved, 24h old after last change and you can merge it. Please just take a moment to create a list of properties that both functions accept so @larkob can update the wiki

**Comment by rogerl50** [comment]
> OK so this PR is approved, 24h old after last change and you can merge it. Please just take a moment to create a list of properties that both functions accept so @larkob can update the wiki

Just checking that the meeting didn't raise any objections to merging this one.

Also, I will draft a new Wiki entry for @larkob.

**Comment by RaphaelAlvez** [comment]
> > OK so this PR is approved, 24h old after last change and you can merge it. Please just take a moment to create a list of properties that both functions accept so @larkob can update the wiki
> 
> Just checking that the meeting didn't raise any objections to merging this one.
> 
> Also, I will draft a new Wiki entry for @larkob.

At the meeting, we couldn't agree if the approach in your PR covers every use case but we agreed to merge it and see what the designers need. Improvements can be added in the future.

**Pull request merged**
